server/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java

```diff
@@ -453,7 +468,12 @@ void executeRequest(
                 }
             }
         });
-        Rewriteable.rewriteAndFetch(original, searchService.getRewriteContext(timeProvider::absoluteStartMillis), rewriteListener);
+
+        Rewriteable.rewriteAndFetch(
+            original,
+            searchService.getRewriteContext(timeProvider::absoluteStartMillis, resolvedLocalIndicesSupplier),
```

Comment:
Can we simply call `resolveLocalIndices` earlier, before calling all of the execute methods, and provide the resolved indices as an argument to them? Indices need to be resolved anyways, I don't follow what the supplier buys us. The resolved indices would then be provided to the rewrite context too.  If that means that other actions like explain and validate query need to be updated accordingly, so be it. Assuming that the requirement of accessing concrete indices when retrieving the rewrite context is clear and won't change, I would recommend making this change separately and directly against main. It's good to focus on that bit only and get it in so that is out of the way.

Reply:
I tried a similar approach and encountered two problems:

- Point in time requests. These requests resolve the indices when the point in time is established and use those resolved indices in all search requests that reference the point in time. When handling such search requests, the concrete indices aren't resolved at all because one or more of the referenced indices could have since been deleted. Instead, `TransportSearchAction` uses the resolved indices from the point in time and allows for partial failures in case of index deletion.
- A small number of query tests expect index resolution to happen after query rewrite logic is executed. Always resolving the indices early breaks those tests.

The first point is the main problem. The previous index resolution approach in `TransportSearchAction` deferred index resolution until after query rewrite logic is executed and made it conditional on a non point-in-time request. This new approach keeps those attributes to prevent regressions, while also creating a resolved local index supplier we can use to lazily generate the index metadata map as necessary.

Reply:
I opened a new PR with only the `QueryRewriteContext` changes against `main`: https://github.com/elastic/elasticsearch/pull/107075